ouroboros-network/src/Ouroboros/Network/Subscription/Common.hs

```diff
@@ -0,0 +1,383 @@
+defaultConnectionAttemptDelay :: DiffTime
+defaultConnectionAttemptDelay = 0.250 -- 250ms delay
+
+-- | Minimum time to wait between connection attempts.
+minConnectionAttemptDelay :: DiffTime
+minConnectionAttemptDelay = 0.010 -- 10ms delay
+
+-- | Maximum time to wait between connection attempts.
+maxConnectionAttemptDelay :: DiffTime
+maxConnectionAttemptDelay = 2 -- 2s delay
+
+-- | Minimum time to wait between ip reconnects
+ipRetryDelay :: DiffTime
+ipRetryDelay = 10 -- 10s delay
+
+sockAddrFamily
+    :: Socket.SockAddr
+    -> Socket.Family
+sockAddrFamily (Socket.SockAddrInet  _ _    ) = Socket.AF_INET
+sockAddrFamily (Socket.SockAddrInet6 _ _ _ _) = Socket.AF_INET6
+sockAddrFamily _                              = Socket.AF_UNSPEC
+
+data ConnectResult = ConnectSuccess
+                   | ConnectSuccessLast
+                   | ConnectFail
+                   deriving (Eq, Ord, Show)
+
+subscribeTo
+    :: HasCallStack
+    => ConnectionTable
+    -> TVar IO (Set ThreadId)
+    -> Tracer IO SubscriptionTrace
+    -> Maybe Socket.SockAddr
+    -> Maybe Socket.SockAddr
+    -> (Socket.SockAddr -> Maybe DiffTime)
+    -> TVar IO Int
+    -> (Socket.Socket -> IO ())
+    -> SubscriptionTarget IO Socket.SockAddr
+    -> IO ()
+subscribeTo tbl threadPool tracer localIPv4_m localIPv6_m connectionAttemptDelay
+  valencyVar k ts = do
+     x <- newTVarM Set.empty
+     doSubscribeTo x ts
+
+  where
+    getLocalAddress :: Socket.SockAddr -> IO (Maybe Socket.SockAddr)
+    getLocalAddress (Socket.SockAddrInet _ _) =
+        case localIPv4_m of
+             Nothing -> do
+                 traceWith tracer $ SubscriptionTraceMissingLocalAddress Socket.AF_INET
+                 return Nothing
+             a  -> return a
+    getLocalAddress (Socket.SockAddrInet6 _ _ _ _) =
+         case localIPv6_m of
+             Nothing -> do
+                 traceWith tracer $ SubscriptionTraceMissingLocalAddress Socket.AF_INET6
+                 return Nothing
+             a  -> return a
+
+    getLocalAddress a = do
+        traceWith tracer $ SubscriptionTraceUnsupportedRemoteAddr a
+        return Nothing
+
+    doSubscribeTo :: TVar IO (Set ThreadId) -> SubscriptionTarget IO Socket.SockAddr -> IO ()
+    doSubscribeTo conThreads targets = do
+        target_m <- getSubscriptionTarget targets
+        case target_m of
+            Just (target, nextTargets) -> do
+                ref <- refConnection tbl target valencyVar
+                case ref of
+                     ConnectionTableCreate -> do
+                        localAddr_m <- getLocalAddress target
+                        case localAddr_m of
+                             Nothing -> -- No usable local address continue with next
+                                        doSubscribeTo conThreads nextTargets
+                             Just localAddr -> do
+                                 -- TODO use asyncWith instead?
```

Comment:
for that to work, every execution path must await for all the threads to not kill them to early; it looks like it's possible.